### DISCUSSION TITLE
[defaultValues] Common practice for object and structure ?

### DISCUSSION THREAD

**Discussion by totomakers** [react-hook-form/react-hook-form#9515] [Q&A]
**Is your feature request related to a problem? Please describe.**

I don't know exactly how we are supposed to manage object & array in defaultValues & i don't have a proper way to provide a good type for defaultValues

**Describe the solution you'd like**

https://codesandbox.io/s/defaultvalues-zoerge?file=/src/App.tsx

I don't know if it's common pattern. But solution like `downshift` provide a select & multiSelect supporting object.
The component we write with this libs except the following input.

For a select:

+ `value:  T | null` <= null is an empty value

For a multiSelect:

+ `value: T[]` <= empty array is the default value

Maybe `DefaultValues` type should be provide in useForm ? 

```
type MyFormDefaultValues = {
  a: SubComponentProps['value']
  b: SubComponentProps['value']
}

const { } = useForm<FormSchema, MyFormDefaultValues>()
```

**Describe alternatives you've considered**

Maybe provide a DefaultValues looking this way ? and provide documentation for override this default type with a `.d.ts` if it's not suit for the project ?

```
array: <= EmptyArray ?
object <= Object | null?
primitive <= Primitive | null?
```

**Comment by bluebill1049** [comment]
thanks for the detailed issue report, I don't think there is an easy to accomplish which you described below:

```
array: <= EmptyArray ?
object <= Object | undefined ?
primitive <= Primitive | undefined ?
```

in terms of `undefined` defaultVlue, that's another point which described in the documentation.


https://react-hook-form.com/api/useform

> Important: You should provide a proper default value and avoid undefined.

> undefined is reserved for fallback from inline defaultValue/defaultChecked to hook level defaultValues.

> undefined value is conflicting with controlled component as default state

**Comment by totomakers** [comment]
Totaly forgot about this undefined rules \o/
Updated using null instead.

**Comment by totomakers** [comment]
Maybe someone with more ts knownlege can make key optionnal if "null" is allowed ?

https://www.typescriptlang.org/play?#code/C4TwDgpgBAsiDKBjAFhAtgQygXigbwCgooIAPDNMAGwgHkAjAKwkWAC58jioA7CiDgGdgAJwCWPAOYBuLgF8ANFzIVqEAIIiRGEBzy9+Q0RMlyA2gF1l5SjQBqGKgFcBUYeKmziK2xAByTmj0ECIcPIHBIrJyBLGgkFAAIhAAZhhOVMAOzhCCADxIqJgkpMAQPAAmglAASiwA9iIVee4mClAYPCAAfN04nMRmANJQElAA1hAg9SlQhegYFhxc3HMoC8MWJWWV1ZraIHmdPVAA-GtFGJtQy6vE85jXZDtVtQ1NLcZS7cd95w9XIZbAA+vAyVBuK24AOuoPCVCoBBicXA0FoTjATmA-WSaQyWUcLnycAB3ViBEQ9R4wig9SxmOA8GQdKoFQAQhBsmIKhx0Qz+oRvDY1PsdBxLEohaoaAEgiEwuDJSVpXQmCx2GCEUqfGpsi4FQikbFKdTsXTgAymXSKlQOVyKgAmXkYrE4QXK3yyyIcACMAAZtcKaKLdFAJdYVQxmKwDVRAyq9a54XGjRSqTTzZbmU4bXbHNyAMzO-nYd06mUReVQf3x3wh8VbWtqKPq2NN+yEpOKpFAA
